dir.h

```diff
@@ -71,6 +71,7 @@ struct pattern_list {
 	 * excludes array above. If non-zero, that check succeeded.
 	 */
 	unsigned use_cone_patterns;
+	unsigned full_cone;
```

Comment:
>However, we instead need to handle the case where that second
line does not exist. Due to the MUSTBEDIR and NEGATIVE flags, the
second pattern becomes "/*".
>
>Create a new option in struct pattern_list to store this "full
cone" mode. When we see the "/*" pattern with no flags, we toggle
full_cone to on. If we later see "/*" pattern with MUSTBEDIR and
NEGATIVE flags, then turn full_cone off.

For my own understanding, are these flags set when the file is parsed?

Reply:
Yes, this is part of the `parse_path_pattern()` method in `dir.c`, which is called by `add_pattern()` inside `add_patterns_from_buffer()`.

Reply:
That method also strips out the tail `/` for directories (`MUSTBEDIR` flag) and the preceding `!` for negations (`NEGATIVE` flag).